File: archive/f/f-sharp/PrimeNumber.fs
open System

module Prime =
    let isPrime n =
        if n < 2 then false
        elif n = 2 then true
        elif n % 2 = 0 then false
        else
            let rec loop d =
                if d * d > n then true
                elif n % d = 0 then false
                else loop (d + 2)
            loop 3

module Helpers =
    let parseArgs =
        function
        | [| input: string |] ->
            match Int32.TryParse input with
            | true, n when n >= 0 -> Ok n
            | _ -> Error "Usage: please input a non-negative integer"
        | _ -> Error "Usage: please input a non-negative integer"

    let handleResult = function
        | Ok true  -> printfn "prime"; 0
        | Ok false -> printfn "composite"; 0
        | Error e  -> eprintfn "%s" e; 1

[<EntryPoint>]
let main argv =
    argv
    |> Helpers.parseArgs
    |> Result.map Prime.isPrime
    |> Helpers.handleResult

File: archive/o/ocaml/prime-number.ml
let has_odd_divisor_under_sqrt n =
  let rec aux k = k <= n / k && (n mod k = 0 || aux (k + 2)) in
  aux 3 (* Assumes n > 3. All other cases will be caught elsewhere *)

let prime = function
  | 0 | 1 -> false
  | 2 | 3 -> true
  | n when n mod 2 = 0 -> false
  | n -> not (has_odd_divisor_under_sqrt n)

let parse_args argv =
  match argv with [| _; n |] -> int_of_string_opt n | _ -> None

let () =
  print_endline
    (match parse_args Sys.argv with
    | Some num when num >= 0 -> if prime num then "Prime" else "Composite"
    | _ -> "Usage: please input a non-negative integer")
